z3c/dependencychecker/tests/test_modules_zcml.py

```diff
@@ -67,172 +72,69 @@ def test_create_from_files_deep_nested(minimal_structure):
-    assert dotted_names == ["my.package.layer"]
-
-
-def test_subscriber_for(tmpdir):
-    dotted_names = _get_zcml_imports_on_file(tmpdir, SUBSCRIBER_FOR)
-    assert len(dotted_names) == 1
-
-
-def test_subscriber_for_details(tmpdir):
-    dotted_names = _get_zcml_imports_on_file(tmpdir, SUBSCRIBER_FOR)
-    assert dotted_names == ["my.package.for"]
-
-
-def test_subscriber_for_multiple(tmpdir):
-    dotted_names = _get_zcml_imports_on_file(tmpdir, SUBSCRIBER_FOR_MULTIPLE)
-    assert len(dotted_names) == 2
-
-
-def test_subscriber_for_multiple_details(tmpdir):
-    dotted_names = _get_zcml_imports_on_file(tmpdir, SUBSCRIBER_FOR_MULTIPLE)
-    assert dotted_names == [
-        "my.package.for",
-        "another.package",
-    ]
-
-
-def test_subscriber_handler(tmpdir):
-    dotted_names = _get_zcml_imports_on_file(tmpdir, SUBSCRIBER_HANDLER)
-    assert len(dotted_names) == 1
-
-
-def test_subscriber_handler_details(tmpdir):
-    dotted_names = _get_zcml_imports_on_file(tmpdir, SUBSCRIBER_HANDLER)
-    assert dotted_names == ["my.package.handler"]
-
-
-def test_securitypolicy_component(tmpdir):
-    dotted_names = _get_zcml_imports_on_file(tmpdir, SECURITYPOLICY_COMPONENT)
-    assert len(dotted_names) == 1
-
-
-def test_securitypolicy_component_details(tmpdir):
-    dotted_names = _get_zcml_imports_on_file(tmpdir, SECURITYPOLICY_COMPONENT)
-    assert dotted_names == ["my.package.component"]
-
-
-def test_gs_registerprofile_provides(tmpdir):
-    dotted_names = _get_zcml_imports_on_file(
-        tmpdir,
-        GS_REGISTERPROFILE_PROVIDES,
-    )
-    assert len(dotted_names) == 1
-
-
-def test_gs_registerprofile_provides_details(tmpdir):
-    dotted_names = _get_zcml_imports_on_file(
-        tmpdir,
-        GS_REGISTERPROFILE_PROVIDES,
-    )
-    assert dotted_names == ["my.package.provides"]
-
-
-def test_ignore_local_import(tmpdir):
-    dotted_names = _get_zcml_imports_on_file(tmpdir, IGNORE_LOCAL_IMPORT)
-    assert len(dotted_names) == 0
-
-
-def test_ignore_asterisk_import(tmpdir):
-    dotted_names = _get_zcml_imports_on_file(tmpdir, ASTERISK)
-    assert len(dotted_names) == 0
+@pytest.mark.parametrize("result,imports", TEST_IMPORTS_MATRIX)
+def test_zcml_include(tmpdir, result, imports):
+    """Check that the ZCML <include directive is scanned."""
+    zcml_stanza = f'<include package="{imports}" />'
+    dotted_names = _get_zcml_imports_on_file(tmpdir, zcml_stanza)
+    _verify_dotted_names(dotted_names, imports, result)
+
+
+@pytest.mark.parametrize("attribute", ("for", "factory", "provides"))
+@pytest.mark.parametrize("result,imports", TEST_IMPORTS_MATRIX)
+def test_zcml_adapter(tmpdir, result, imports, attribute):
+    """Check that the ZCML <adapter directive is scanned."""
+    zcml_stanza = f'<adapter {attribute}="{imports}" />'
+    dotted_names = _get_zcml_imports_on_file(tmpdir, zcml_stanza)
+    _verify_dotted_names(dotted_names, imports, result)
```

Comment:
2x parametrize decorator *and* f-strings using those attributes: fancy! Nice solution.